private/bufpkg/bufimage/bufimagemodify/internal/internal.go

```diff
@@ -83,7 +83,25 @@ func RemoveLocationsFromSourceCodeInfo(sourceCodeInfo *descriptorpb.SourceCodeIn
 	// are structured as expected, and collect all of the indices that
 	// we need to delete.
 	indices := make(map[int]struct{}, len(paths)*2)
+	potentialFieldOptionToChildrenIndices := make(map[string][]int, len(paths))
+	// If multiple paths are the same, use the last one. This is fine because
+	// this is only useful for field options, where each FieldOptions' path does
+	// not appear twice.
+	keyToIndex := make(map[string]int, len(paths))
 	for i, location := range sourceCodeInfo.Location {
+		path := location.Path
+		keyToIndex[GetPathKey(path)] = i
+		if isPathMaybeForFieldOption(path) {
+			parentKey := GetPathKey(path[:len(path)-1])
```

Comment:
If the source file used compound option names or repeated options, the length of the path could be greater than the path to options +1.

For example:
```
(custom.option.name).subfield_name.another_subfield_name = "abc"
```
In the above, this path will have three extra components beyond the path to the options message: one for the the `custom.option.name` tag, another for `subfield_name`, and a final one for `another_subfield_name`.

Similarly:
```
(custom.repeated.field) = 123
```
In the above, the path will have two elements beyond the options message: the tag for `custom.repeated.field` plus an index.

I think you may want to invert how you are identifying these: identify the "roots" as all paths to field options messages and store into a prefix tree instead of a map. With a prefix tree, you can then easily test subsequent paths to see if they are a sub-path of a root. When you find such a sub-path for an element that is _not_ being deleted, increment a count for that root. At the end, all roots in the prefix tree with a count of zero should be removed as well since they have no retained descendants.

BTW, I guess `isPathMaybeForFieldOption` already exists as I don't see the source in this PR. But this can be made definitive (instead of "maybe") as it's a fairly simple DFA to match any option. Similarly, you can definitively match a field options _message_ (the root path for any field option). The DFA for that looks like so, which should be straight-forward to encode as a static structure:
```mermaid
flowchart TD;
  Start -- "7 (tag extension)" --> fields;
  fields --"* (index)" --> field;
  Start -- "4 (tag message_type)" --> messages;
  messages -- "* (index)" --> message;
  message -- "3 (tag nested_type)" --> messages;
  message -- "2 (tag field)" --> fields;
  message -- "6 (tag extension)" --> fields;
  field -- "8 (tag options)" --> Accept;
```